### DISCUSSION TITLE
How to access data from Array inside Array with API method

### DISCUSSION THREAD

**Discussion by Winniepoom** [Budibase/budibase#9758] [Help]
Hello I had a n question I try to access data from Redmine and It look like it had array inside array.
Examples: `GET /projects/1/memberships.xml`
Response: 
```
<?xml version="1.0" encoding="UTF-8"?>
<memberships type="array" limit="25" offset="0" total_count="3">
  <membership>
    <id>1</id>
    <project name="Redmine" id="1"/>
    <user name="David Robert" id="17"/>
    <roles type="array">
      <role name="Manager" id="1"/>
    </roles>
  </membership>
  <membership>
    <id>3</id>
    <project name="Redmine" id="1"/>
    <group name="Contributors" id="24"/>
    <roles type="array">
      <role name="Contributor" id="3"/>
    </roles>
  </membership>
  <membership>
    <id>4</id>
    <project name="Redmine" id="1"/>
    <user name="John Smith" id="27"/>
    <roles type="array">
      <role name="Developer" id="2" />
      <role name="Contributor" id="3" inherited="true" />
    </roles>
  </membership>
</memberships>
```
I already try to transformer data `return data.membership`  but it still had array inside Result after Transformer look like this.
```
{"$":{"id":"2","name":"Redmine"}}
{"$":{"id":"6","name":"James"}}
{"$":{"type":"array"},"role":{"$":{"id":"3",
``` 

Question Is how do I get data inside that array Sorry I'm kind of newbie in this field.
Thank you

**Comment by andz-bb** [comment]
hey @Winniepoom, can you give an example of what sort of data structure you would like once the data has been transformed? You can definitely pull the data from the array out and return it in a flatter way with the transformer, I would just need to know how you would want it to be presented.

**Reply by Winniepoom** [reply]
@andz-bb 
Bump.

**Reply by andz-bb** [reply]
hey @Winniepoom 
you would need to map the response object to deal with the roles array. It's an array because some of the users have more than one role. The best way to deal with this is to take each role name from the array and join them, replacing the array with a string of all roles.

How the XML is parsed into an object makes it a little bit more awkward to deal with but hopefully the code below for your transformer makes sense.

```js
return data.membership.map(mem => {
  // do this for each membership object

  const { role } = mem.roles;
  // if the role is an array join it into a string
  const roles = Array.isArray(role) ? role.map(role => role.$.name).join(', ') : role.$.name; 

  // return a new membership object with our new roles string replacing the array
  return { ...mem, roles };
});
```
<img width="1089" alt="xml response transformer" src="https://user-images.githubusercontent.com/110921612/220666235-ca1e2b9b-edb9-4aaf-ac3f-1e0ebd542dcb.png">

Here's one of the JSON objects after the transformer has run:

```json
{
  "id": "5849",
  "project": {
    "$": {
      "name": "Redmine",
      "id": "1"
    }
  },
  "user": {
    "$": {
      "name": "Jon Smyth",
      "id": "2723234"
    }
  },
  "roles": "Developer, Contributor"
}
```

You can use the same principle to flatten the other objects and make your API response more usable in Budibase. Hope this helps!

**Comment by Winniepoom** [comment]
This data is a membership of project in Redmine. So I want it to be present like Their ID in Redmine, Project Name, Project ID, Username and Roles . For Example according to Example API Respond.
ID = 1.
Project Name = Redmine.
Project ID = 1.
Username =David Robert.
User ID = 17.
Role Name = Manager.

This is I would like to get.
Thank you.
